cg/smart_contracts/ClearGDPR.sol

```diff
@@ -1,135 +1,152 @@
-    return subjects[_subjectIdHash].isErased;
-  }
+    function getIsErased(bytes32 _subjectId) public view returns(bool){
+        return subjects[_subjectId].isErased;
+    }
 
-  function getSubjectDataState(bytes32 _subjectIdHash, address _processorIdHash) public view returns(State){
-    return subjects[_subjectIdHash].statePerProcessor[_processorIdHash];
-  }
+    function getSubjectDataState(bytes32 _subjectId, address _processor) public view returns(State){
+        return subjects[_subjectId].statePerProcessor[_processor];
+    }
 
-  function getProcessors() public view returns(address[], uint256){
-    return (processors, processors.length);
-  }
+    function getProcessors() public view returns(address[], uint256){
+        return (processors, processors.length);
+    }
 
-  function setSubjectDataState(bytes32 _subjectIdHash, address _processorIdHash, State _state) public returns(bool){
-    subjects[_subjectIdHash].statePerProcessor[_processorIdHash] = _state;
-    return true;
-  }
+    function setSubjectDataState(bytes32 _subjectId, address _processor, State _state) public returns(bool){
+        subjects[_subjectId].statePerProcessor[_processor] = _state;
+        return true;
+    }
 
-  function setProcessors(address[] _newProcessors) public returns(bool){ //BUG IN HERE, CAN'T USE BYTES32
-    processors = new address[](_newProcessors.length);
-    processors = _newProcessors;
-    return true;
-  }
-  
-  function isProcessor(address _processor) public view returns(bool){ //0x1 for remix, since it's the only argument
-    for(uint256 i = 0; i < processors.length; i++){
-      if(processors[i] == _processor){
+    function setProcessors(address[] _newProcessors) public returns(bool){
+        processors = new address[](_newProcessors.length + 1);
+        processors[0] = controller;
+        for(uint256 i = 0; i < _newProcessors.length; i++){
+            require(_newProcessors[i] != controller);
+            for(uint256 j = i + 1; j < _newProcessors.length; j++){ // Check if the processors are unique
+                require(_newProcessors[i] != _newProcessors[j]);
+            }
+            processors[i + 1] = _newProcessors[i];
+        }
         return true;
-      }
+    } 
+    //processors = _newProcessors; //Interesting line
+  
+    function isProcessor(address _processor) public view returns(bool){
+        for(uint256 i = 0; i < processors.length; i++){
+            if(processors[i] == _processor){
+                return true;
+            }
+        }
+        return false;
     }
-    return false;
-  }
   
-  function areAllValidProcessors(address[] _processors) public view returns(bool){
-    for(uint256 i = 0; i < _processors.length; i++){
-      require(isProcessor(_processors[i]));
+    function areAllValidProcessors(address[] _processors) public view returns(bool){
+        for(uint256 i = 0; i < _processors.length; i++){
+            require(isProcessor(_processors[i]));
```

Comment:
Does it automatically return false if require fails?

Reply:
No, it reverts anything that was made up to that point in the function, to the state prior to the function call. The return true is to quickly test if the function executed successfully, even if the behavior was not correct